### ISSUE TITLE
Allow to use generics as return type for setters when chaining

### ISSUE THREAD

**Issue by mkurz** [rzwitserloot/lombok#1114]
When using `lombok.accessors.chain = true` this code...

```java
public class GetterSetterExample {
	@Setter
	private int age;
}
```
...looks like this in Vanilla Java:
```java
public class GetterSetterExample {

	private int age;

	public GetterSetterExample setAge(int age) {
		this.age = age;
		return this;
	}
}
```

Instead we would like to be able to generate a setter like this one:

```java
public class GetterSetterExample {

	private int age;

	public <T extends GetterSetterExample> T setAge(int age) {
		this.age = age;
		return (T)this;
	}
}
```
This would make chaining with subclasses more comfortable.
Is this (somehow) possible already? If not - how hard would it be to add this feature?

**Comment by randall-mo** [comment]
I would like to be able generate a setter like:
`public  class Base<T extends Base<T>> {
        public T set(){
            return (T)this;
        }
    }
    
    public class A<T extends A<T>> extends Base<T>{
        public T setA(){
            return (T)this;
        }
    }
    
    public  class B<T extends B<T>> extends A<T>{
        public  T setB(){
            return (T)this;
        }
    }`

**Comment by GedMarc** [comment]
+1 Please,

CRP Pattern :)

**Comment by GedMarc** [comment]
@Accessors(chain=true)
@Setter(returnType="T")?

**Comment by Maaartinus** [comment]
@mkurz 

> This would make chaining with subclasses more comfortable.

Yes, but also inherently unsafe.
```
class Base {
    public <T extends Base> T setAge(int age) {
        this.age = age;
        return (T)this;
    }
}

class A extends Base {}
class B extends Base {}

A a = new B().setAge(10); // BANG!
```

> Is this (somehow) possible already?

No.

> If not - how hard would it be to add this feature?

I guess, in the form you proposed, it should be simple, but given how dangerous it is, I don't think, it's a good idea.

An alternative would be to automatically override the setters in the subclasses, but that would lead to a code bloat and it's hard to implement (as you need information about super fields before generating the methods, see https://github.com/rzwitserloot/lombok/wiki/LOMBOK-CONCEPT:-Resolution).

**Comment by GedMarc** [comment]
Yes it causes a lot of code bloat overriding the setters to return properly,

This is the *only* solution to the inheritance resolving self problem though, the unchecked cast of 
```
return (T)this;
```
no longer causes a warning on JDK 9 and up as well as it gets resolved properly.
This is definitely a needed request for modular, and domain driven applications built in Java.
It is the only missing item that does not allow me to use Lombok.

**Comment by AndreKoepke** [comment]
I'm also looking fot this feature. Is that easier, when we define something like this:

```java
@Setter(genericType = true)
public class Parent<T> {
    private int age;
}

// and this class looks delomboked like this:

public class Parent<T> {

    private int age;

    public T setAge(int age) {
        this.age = age;
        return (T) this;
    }
}

What lombok has to do:
1. Switch return-type to `T`, if `genericType` is true.
2. And type-cast to `T`, if `genericType` is true.

I'm not familiar with the lombok internals, but it sounds easy to me.
Or is there a special need for `<T extends Parent>` in the method-type?
